File: src/main/java/xyz/gnarbot/gnar/commands/general/GoogleCommand.java
package xyz.gnarbot.gnar.commands.general;

import net.dv8tion.jda.entities.Message;
import org.apache.commons.lang3.StringUtils;
import org.jsoup.Jsoup;
import org.jsoup.nodes.Element;
import org.jsoup.select.Elements;
import xyz.gnarbot.gnar.handlers.commands.Command;
import xyz.gnarbot.gnar.handlers.commands.CommandExecutor;
import xyz.gnarbot.gnar.utils.Note;

import java.io.IOException;
import java.net.URLDecoder;
import java.net.URLEncoder;
import java.nio.charset.StandardCharsets;
import java.util.StringJoiner;

@Command(
        aliases = "google",
        usage = "(query)",
        description = "Who needs browsers!?"
)
public class GoogleCommand extends CommandExecutor
{
    @Override
    public void execute(Note note, String label, String[] args)
    {
        if (args.length == 0)
        {
            note.reply("Gotta have a query to Google.");
            return;
        }
        
        try
        {
            String query = StringUtils.join(args, " ");
            Message msg = note.reply("Searching `" + query + "`.");
            
            String userAgent = "GN4R-Bot";
            
            Elements links = Jsoup.connect(
                    String.format("http://www.google.com/search?q=%s", URLEncoder.encode(query, StandardCharsets.UTF_8.displayName())))
                    .userAgent(userAgent).get().select(".g>.r>a");
            
            StringJoiner joiner = new StringJoiner("\n");
            
            for (Element link : links)
            {
                String title = link.text();
                String url = link.absUrl("href");
                url = URLDecoder.decode(url.substring(url.indexOf('=') + 1, url.indexOf('&')), StandardCharsets.UTF_8.displayName());
                
                if (!url.startsWith("http"))
                {
                    continue;
                }
                
                joiner.add("Title: **" + title + "**");
                joiner.add("URL: " + url);
                
                break;
            }
            
            
            if (!links.isEmpty()) msg.updateMessage(joiner.toString());
            else
                msg.updateMessage(String.format("%s ➜ No results for `%s`.", note.getAuthor().getAsMention(), query));
        }
        catch (IOException e)
        {
            note.reply("Caught an exception while trying to Google stuff.");
            e.printStackTrace();
        }
    }
}

File: src/main/kotlin/xyz/gnarbot/gnar/commands/executors/general/GoogleCommand.kt
package xyz.gnarbot.gnar.commands.executors.general

import org.jsoup.Jsoup
import xyz.gnarbot.gnar.Constants
import xyz.gnarbot.gnar.commands.handlers.Command
import xyz.gnarbot.gnar.commands.handlers.CommandExecutor
import xyz.gnarbot.gnar.utils.Note
import java.io.IOException
import java.net.URLEncoder
import java.nio.charset.StandardCharsets

@Command(aliases = arrayOf("google"), usage = "-query...", description = "Who needs browsers!?")
class GoogleCommand : CommandExecutor() {
    override fun execute(note: Note, args: List<String>) {
        if (args.isEmpty()) {
            note.respond().error("Gotta have something to search Google.").queue()
            return
        }

        try {
            val query = args.joinToString(" ")

            val blocks = Jsoup.connect("http://www.google.com/search?q=${URLEncoder.encode(query, StandardCharsets.UTF_8.displayName())}")
                    .userAgent("Gnar")
                    .get()
                    .select(".g")

            if (blocks.isEmpty()) {
                note.respond().error("No search results for `$query`.").queue()
                return
            }

            note.respond().embed {
                color = Constants.COLOR
                setAuthor("Google Results", "https://www.google.com/", "https://www.google.com/favicon.ico")
                thumbnail = "https://gnarbot.xyz/assets/img/google.png"

                description {
                    var count = 0

                    for (block in blocks) {
                        if (count >= 3) break

                        val list = block.select(".r>a")
                        if (list.isEmpty()) break

                        val entry = list[0]
                        val title = entry.text()
                        val url = entry.absUrl("href").replace(")", "\\)")
                        var desc: String? = null

                        val st = block.select(".st")
                        if (!st.isEmpty()) desc = st[0].text()

                        appendln(b(link(title, url))).appendln(desc)

                        count++
                    }
                }
            }.rest().queue()
        } catch (e: IOException) {
            note.respond().error("Caught an exception while trying to Google stuff.").queue()
            e.printStackTrace()
        }
    }
}
